src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,298 @@
+          refer to cached config file. Is supplied by the parent
+          ConfigParser.
+        _nest_path: The names of the ConfigParsers that this option is
+          nested in. Used be able to efficiently resolve where to grab
+          value out of configuration file and construct environment
+          variable name. This is supplied by the parent ConfigParser.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        parse_str: Callable[[str], _T] | None = None,
+        choices: Iterable[Any] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: Name to assign to this ConfigOption.
+            parse_str: String parser function for this ConfigOption.
+            choices: List of possible values for this ConfigOption.
+            env_name: Environmental variable name value should be read from.
+            _root_parser: Reference to the root parser. Should be supplied by
+              the parent ConfigParser.
+            _nest_path: The names of the ConfigParsers that this option is
+              nested in. This is supplied by the parent ConfigParser.
+        """
+        self.name = name
+        self.parse_str = parse_str
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self) -> Any:
+        """Retrieve a value of option.
+
+        This function implements order of precedence between different sources.
+        """
+        source = "environment variable"
+        loaded_env, value = self._get_env()
+        if not loaded_env:
+            source = "configuration file"
+            value = self._get_config()
+        if self.choices and value not in self.choices:
+            raise ConfigSourceError(
+                f"The value of {self.option_name} read from "
+                f"{source} is not part of {self.choices}"
+            )
+        return value
+
+    @property
+    def option_name(self) -> str:
+        """User-friendly name of the config option. Includes self._nest_path."""
+        return ".".join(self._nest_path[1:])
+
+    @property
+    def default_env_name(self) -> str:
+        """The default environmental variable name for this option."""
+        pieces = map(lambda e: e.upper(), self._nest_path[1:])
+        return f"SNOWFLAKE_{'_'.join(pieces)}"
+
+    def _get_env(self) -> tuple[bool, str | _T | None]:
+        """Get value from environment variable if possible.
+
+        Returns whether it was able to load the data and the loaded value
+        itself.
+        """
+        if self.env_name is False:
+            return False, None
+        if self.env_name is not None:
```

Comment:
Nit: if/else could be shortened to `self._env_name = self._env_name if self.env_name is not None else self.default_env_name`.

Reply:
The only difference is that my comment won't really fit anywhere nicely. Should we sacrifice it for being a one-liner?
Could you maybe add a suggestion edit so that we could all visualize it first?

Reply:
```
 # Generate environment name if it was not explicitly supplied,
 #  and isn't disabled
 self._env_name = self._env_name if self.env_name is not None else self.default_env_name
```

I may be even ok with excluding empty string as well
```
 # Generate environment name if it was not explicitly supplied,
 #  and isn't disabled
 self._env_name = self._env_name if self.env_name else self.default_env_name
```

Reply:
I'm cool with leaving it to @sfc-gh-mkeller to decide.  I'll just point out that Guido chose ternary operator syntax deliberately so it would be a little bit ugly and people wouldn't over use it. 😄 .  There's nothing inherently wrong with if-else blocks.